### PR TITLE
Remove monkey patches of `Range#each` and `Range#step`

### PR DESCRIPTION
**Pull request by lovro-bikic**
Reverts https://github.com/rails/rails/pull/11474.

#### Range#each

When the first range element doesn't implement `#succ`, [`Range#each` raises a TypeError](https://rubyapi.org/3.3/o/range#method-i-each). The monkey patch implements the same behavior as Ruby, so it adds no value.

#### Range#step

[Ruby 3.4 will introduce a behavior change to `Range#step`](https://github.com/ruby/ruby/pull/7444) which will produce sensible results for some non-numeric types since it will use the `#+` method to generate successive elements. This change would make code such at this possible (tested on version `3.4.0-preview2`):
```ruby
(Time.current..).step(30.minutes).first(3)
# => 
# [Thu, 12 Dec 2024 23:55:41.568749000 CET +01:00,
#  Fri, 13 Dec 2024 00:25:41.568749000 CET +01:00,
#  Fri, 13 Dec 2024 00:55:41.568749000 CET +01:00]
```
since [`ActiveSupport::TimeWithZone` implements its own `#+`](https://github.com/rails/rails/blob/ef4201c204e6ad37537d76cc471de27a8fd53d32/activesupport/lib/active_support/time_with_zone.rb#L298-L314).

However, this doesn't work currently because `Range#step` has been overridden and it specifically prevents an `ActiveSupport::TimeWithZone` boundary:
```ruby
(Time.current..).step(30.minutes).first(3)
# can't iterate from ActiveSupport::TimeWithZone (TypeError)
```

That override was introduced in https://github.com/rails/rails/pull/11474 at a time when Ruby's behavior was different.

For versions < 3.4, `Range#step` will raise a TypeError like `Range#each`, so the monkey patch adds no value for those versions.

### Detail

This PR removes monkey patches of `Range#each` and `Range#step`.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activesupport/lib/active_support/core_ext/range/each.rb
Comment (lovro-bikic) [inline activesupport/lib/active_support/core_ext/range/each.rb:1]: I think it's also worth removing `Range#each` since [Ruby already raises the same error](https://github.com/ruby/ruby/blob/5e8c9b4b284d3200dc74ff41243288caf0d7b4f6/spec/ruby/core/range/size_spec.rb#L78-L94) even without the custom implementation:
```ruby
(Time.current..).size
# can't iterate from ActiveSupport::TimeWithZone (TypeError)
```

Not sure though if this should be done in a separate PR.

**Comment by matthewd** [comment]
> Ruby 3.4 will introduce

> Ruby already raises

What behaviour do we want on other existing and supported Ruby versions?

**Comment by lovro-bikic** [comment]
> What behaviour do we want on other existing and supported Ruby versions?

Good point, forgot to put this in. Older Ruby versions already raise the `can't iterate from X` error for both `Range#each` and `Range#step`, so if we remove the override, the same behavior will be preserved. Essentially, `EachTimeWithZone` module only reimplements what Ruby already does.

If I [run the example](https://github.com/rails/rails/pull/11474#issue-16858306) from the PR where this was introduced without the special handling, I get the same error from Ruby:
```ruby
((Time.now.in_time_zone - 1.month)..Time.now.in_time_zone).to_a
# can't iterate from ActiveSupport::TimeWithZone (TypeError)
```

Some [additional reference](https://rubyapi.org/3.3/o/range#method-i-each) from Ruby docs:
> Raises an exception unless self.first.respond_to?(:succ).

`TimeWithZone` doesn't implement `#succ`:
```ruby
Time.current.respond_to?(:succ)
=> false
```

File: activesupport/test/core_ext/range_ext_test.rb [hunk 1]
```diff
@@ -253,13 +253,6 @@ def test_each_on_time_with_zone
     end
   end
 
-  def test_step_on_time_with_zone
-    twz = ActiveSupport::TimeWithZone.new(nil, ActiveSupport::TimeZone["Eastern Time (US & Canada)"], Time.utc(2006, 11, 28, 10, 30))
-    assert_raises TypeError do
-      ((twz - 1.hour)..twz).step(1) { }
-    end
```
Comment (byroot) [inline activesupport/test/core_ext/range_ext_test.rb:260]: I think we should keep that test (or a similar one) but with a ruby version check. `if RUBY_VERSION < "3.4"`, and then add another test to make sure it works as expected on 3.4+

File: activesupport/lib/active_support/core_ext/range/each.rb
Comment (byroot) [reply activesupport/lib/active_support/core_ext/range/each.rb:1]: You can do both in one.

**Comment by lovro-bikic** [comment]
@byroot I've updated the whole PR (title, description, commit) so it removes `Range#each` in addition to `Range#step`, and I've added a test for `Range#step` for Ruby >= 3.4.

**Review by byroot — APPROVED** [review]
